File: py/packages/genkit/src/genkit/_core/_instrumentation.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Telemetry dispatcher. Composes Instrumentation providers; no OpenTelemetry."""

from __future__ import annotations

import contextlib
import inspect
from collections.abc import Awaitable, Callable, Mapping
from contextvars import ContextVar
from typing import TypeVar

from ._instrumentation_api import Instrumentation, SpanContext, SpanMetadata

T = TypeVar('T')

instrumentations: list[Instrumentation] = []

# Active SpanContext so set_custom_metadata_attributes can reach it.
current_span: ContextVar[SpanContext | None] = ContextVar('genkit_span_context', default=None)


def describe_value(value: object) -> str:
    """Module-qualified name of a type or of an instance's type."""
    if isinstance(value, type):
        return f'type {value.__module__}.{value.__qualname__}'
    cls = type(value)
    return f'{cls.__module__}.{cls.__qualname__}'


def configure_instrumentation(instrumentation: Instrumentation) -> None:
    """Turn on a telemetry backend. Call before ``Genkit()`` to stack backends.

    Each provider wraps the next. ``OtelInstrumentation`` is the built-in.
    ``genkit start`` installs that one when a collector is configured.
    """
    # The class itself has run_in_new_span, so a forgotten () would pass a
    # Protocol check and then die inside OTel on the first span.
    if isinstance(instrumentation, type) or not isinstance(instrumentation, Instrumentation):
        raise TypeError(
            'configure_instrumentation expected an Instrumentation instance, got ' + describe_value(instrumentation)
        )
    instrumentations.append(instrumentation)


def reset_instrumentation() -> None:
    """Remove all providers. Tests and re-init."""
    instrumentations.clear()
    from ._otel_instrumentation import reset_developer_ui_collector

    reset_developer_ui_collector()


def is_instrumented_by(kind: type) -> bool:
    """True when a configured provider is an instance of ``kind``.

    Use ``is_instrumented_by(OtelInstrumentation)`` to see whether Genkit
    is already minting OpenTelemetry spans.
    """
    if not isinstance(kind, type):
        raise TypeError('is_instrumented_by expected a type, got ' + describe_value(kind))
    return any(isinstance(i, kind) for i in instrumentations)


def set_custom_metadata_attributes(attributes: Mapping[str, object]) -> None:
    """Write metadata on the active span. No-op outside a span."""
    span = current_span.get()
    if span is not None:
        span.set_metadata(attributes)


async def run_in_new_span(
    name: str,
    fn: Callable[[SpanContext], Awaitable[T]],
    *,
    action_type: str | None = None,
    input: object | None = None,
    attributes: Mapping[str, str] | None = None,
    subtype: str | None = None,
    init: object | None = None,
    metadata: Mapping[str, object] | None = None,
    is_root: bool | None = None,
) -> T:
    """Run ``fn`` inside a new span via the configured provider chain.

    No providers → ``fn`` runs with a no-op span (empty ids). Index 0 is
    outermost. The provider list is snapshotted so configure/reset during an
    await cannot break the chain.
    """
    if not inspect.iscoroutinefunction(fn):
        name = getattr(fn, '__qualname__', type(fn).__name__)
        raise TypeError(f'run_in_new_span expected an async callback, got {name}')
    meta = SpanMetadata(
        name=name,
        action_type=action_type,
        input=input,
        attributes=dict(attributes) if attributes else {},
        subtype=subtype,
        init=init,
        metadata=metadata,
        is_root=is_root,
    )
    providers = list(instrumentations)
    if not providers:
        return await _run_with_span(NoopSpanContext(), fn)

    spans: list[SpanContext] = []

    async def build(index: int) -> T:
        if index == len(providers):
            return await _run_with_span(CompositeSpanContext(spans), fn)

        async def nxt(span: SpanContext) -> T:
            spans.append(span)
            return await build(index + 1)

        return await providers[index].run_in_new_span(meta, nxt)

    return await build(0)


async def _run_with_span(
    span: SpanContext,
    fn: Callable[[SpanContext], Awaitable[T]],
) -> T:
    token = current_span.set(span)
    try:
        return await fn(span)
    finally:
        current_span.reset(token)


class NoopSpanContext:
    """Span used when nothing is configured."""

    @property
    def trace_id(self) -> str:
        return ''

    @property
    def span_id(self) -> str:
        return ''

    def set_metadata(self, metadata: Mapping[str, object]) -> None:
        return

    def set_output(self, value: object) -> None:
        return


class CompositeSpanContext:
    """Fans metadata/output to every provider; ids are first non-empty."""

    def __init__(self, spans: list[SpanContext]) -> None:
        self._spans = spans

    @property
    def trace_id(self) -> str:
        return self._first_non_empty(lambda s: s.trace_id)

    @property
    def span_id(self) -> str:
        return self._first_non_empty(lambda s: s.span_id)

    def set_metadata(self, metadata: Mapping[str, object]) -> None:
        for span in self._spans:
            with contextlib.suppress(Exception):
                span.set_metadata(metadata)

    def set_output(self, value: object) -> None:
        for span in self._spans:
            with contextlib.suppress(Exception):
                span.set_output(value)

    def _first_non_empty(self, get: Callable[[SpanContext], str]) -> str:
        for span in self._spans:
            value = get(span)
            if value:
                return value
        return ''

File: go/plugins/otel/instrumentation.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

// Package otel provides GenAiInstrumentation, an OpenTelemetry GenAI
// semantic-conventions provider for Genkit.
//
// The application owns SDK setup: configure a TracerProvider / MeterProvider /
// LoggerProvider (e.g. via the OTel SDK) before constructing Genkit. When no
// provider is configured, go.opentelemetry.io/otel returns non-recording spans
// and no-op instruments, so this provider is effectively inert.
//
// Wire it up with [tracing.ConfigureInstrumentation]:
//
//	tracing.ConfigureInstrumentation(otel.NewGenAiInstrumentation(otel.GenAiInstrumentationOptions{
//		ContentCapturingMode: genai.SpanOnly,
//		EmitToolSpans:        true,
//	}))
//	g := genkit.Init(ctx, genkit.WithPlugins(&googlegenai.GoogleAI{}))
//
// Configuration is process-wide on purpose: Genkit actions (e.g. a model
// obtained straight from a plugin) can run without a Genkit instance, and they
// must be instrumented too.
//
// The provider replaces Genkit's default OpenTelemetry encoding (the genkit:*
// span attributes) and composes with the built-in dev instrumentation, which
// feeds the Developer UI on a separate pipeline. Do not combine it with the
// googlecloud or firebase telemetry plugins: those read the genkit:* attributes
// (their metrics stop) and their redaction does not cover gen_ai.* content.
//
// See the spec:
// https://github.com/open-telemetry/semantic-conventions-genai
package otel

import (
	"context"
	"log/slog"
	"sync"

	"go.opentelemetry.io/otel"
	otellog "go.opentelemetry.io/otel/log"
	otellogglobal "go.opentelemetry.io/otel/log/global"
	"go.opentelemetry.io/otel/metric"
	oteltrace "go.opentelemetry.io/otel/trace"

	"github.com/firebase/genkit/go/core/api"
	"github.com/firebase/genkit/go/core/tracing"
	"github.com/firebase/genkit/go/internal"
	"github.com/firebase/genkit/go/plugins/otel/genai"
)

// GenAiInstrumentationOptions configures a [GenAiInstrumentation].
type GenAiInstrumentationOptions struct {
	// ContentCapturingMode selects where spec-shaped GenAI message content
	// (gen_ai.system_instructions, gen_ai.input.messages, gen_ai.output.messages)
	// is recorded. Content may contain PII, so the default is NoContent. When
	// the zero value is passed, the env var CaptureContentEnvVar is consulted;
	// an explicit value here overrides it.
	ContentCapturingMode genai.ContentCapturingMode

	// CaptureActionIO records raw Genkit action input/output as genkit.input /
	// genkit.output JSON attributes on every span (model, tool, flow, etc.).
	// Independent of ContentCapturingMode. May contain PII, off by default.
	CaptureActionIO bool

	// EmitToolSpans emits execute_tool spans for tool actions. Off by default.
	EmitToolSpans bool

	// DisableMetrics turns off the spec's GenAI client metrics (token usage,
	// operation duration). Metrics are on by default (the zero value): low
	// cardinality and cheap.
	DisableMetrics bool

	// ScopeName is the instrumentation scope name for the tracer/logger/meter.
	// Defaults to "genkit-genai".
	ScopeName string

	// Tracer optionally injects a tracer (escape hatch). When nil the tracer is
	// resolved from the global TracerProvider.
	Tracer oteltrace.Tracer

	// Meter optionally injects a meter (escape hatch). When nil the meter is
	// resolved from the global MeterProvider.
	Meter metric.Meter
}

// GenAiInstrumentation is a [tracing.Instrumentation] that emits OpenTelemetry
// telemetry following the OTel GenAI semantic conventions.
type GenAiInstrumentation struct {
	contentMode     genai.ContentCapturingMode
	captureOnSpan   bool
	captureOnEvent  bool
	captureActionIO bool
	emitToolSpans   bool

	tracer oteltrace.Tracer
	logger otellog.Logger
	// metrics is nil when disabled or when instrument creation failed.
	metrics *genai.Metrics

	warnOnce sync.Once
}

// NewGenAiInstrumentation builds a GenAiInstrumentation from options.
//
// The tracer, meter, and logger are resolved here, once. That still picks up
// an SDK installed later: the OTel globals hand out delegating instruments
// that switch to the real provider when it is set. Resolving per span instead
// would take the SDK provider's lock on every span start.
func NewGenAiInstrumentation(opts GenAiInstrumentationOptions) *GenAiInstrumentation {
	mode := opts.ContentCapturingMode
	if mode == "" {
		mode = contentCapturingModeFromEnv()
	}
	scope := opts.ScopeName
	if scope == "" {
		scope = "genkit-genai"
	}
	g := &GenAiInstrumentation{
		contentMode:     mode,
		captureOnSpan:   mode == genai.SpanOnly || mode == genai.SpanAndEvent,
		captureOnEvent:  mode == genai.EventOnly || mode == genai.SpanAndEvent,
		captureActionIO: opts.CaptureActionIO,
		emitToolSpans:   opts.EmitToolSpans,
		tracer:          opts.Tracer,
		logger: otellogglobal.Logger(scope,
			otellog.WithInstrumentationVersion(internal.Version),
			otellog.WithSchemaURL(genai.SchemaURL)),
	}
	if g.tracer == nil {
		g.tracer = otel.Tracer(scope,
			oteltrace.WithInstrumentationVersion(internal.Version),
			oteltrace.WithSchemaURL(genai.SchemaURL))
	}
	if !opts.DisableMetrics {
		meter := opts.Meter
		if meter == nil {
			meter = otel.Meter(scope,
				metric.WithInstrumentationVersion(internal.Version),
				metric.WithSchemaURL(genai.SchemaURL))
		}
		m, err := genai.NewMetrics(meter)
		if err != nil {
			slog.Warn("failed to create genai metrics, metrics disabled", "error", err)
		} else {
			g.metrics = m
		}
	}
	return g
}

// contentCapturingModeFromEnv reads the spec env var. It runs at construction
// with no context, hence slog rather than core/logger.
func contentCapturingModeFromEnv() genai.ContentCapturingMode {
	mode, ok := genai.ParseContentCapturingMode(getenv(genai.CaptureContentEnvVar))
	if !ok {
		slog.Warn("invalid content capturing mode, defaulting to NO_CONTENT",
			"env", genai.CaptureContentEnvVar)
	}
	return mode
}

// RunInNewSpan dispatches on the Genkit action type and encodes the span using
// the OTel GenAI conventions.
func (g *GenAiInstrumentation) RunInNewSpan(ctx context.Context, info *tracing.SpanInfo, next tracing.NextFunc) (any, error) {
	if info == nil {
		// Nothing to encode; keep the chain intact.
		return next(ctx, nil)
	}
	actionType := info.Subtype()
	if actionType == "" {
		actionType = info.Type()
	}
	switch actionType {
	case string(api.ActionTypeModel):
		return g.runModelSpan(ctx, info, next)
	case string(api.ActionTypeTool), string(api.ActionTypeToolV2):
		// DefineTool registers tool.v2; plain "tool" is the legacy spelling.
		if g.emitToolSpans {
			return g.runToolSpan(ctx, info, next)
		}
	}
	return g.runGenericSpan(ctx, info, next, actionType)
}

var _ tracing.Instrumentation = (*GenAiInstrumentation)(nil)
